src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,164 @@
+  private double mShooterVelocitySetpoint = 0.0;
+  
+  private double mP = 0.0;
+  private double mI = 0.0;
+  private double mD = 0.0;
+
+  private double mGreenShootingVelocity = 0.0;
+  private double mYellowShootingVelocity = 0.0;
+  private double mBlueShootingVelocity = 0.0;
+  private double mRedShootingVelocity = 0.0;
+
+  public ShooterSubsystem() {
+    mShooterMotor = new MotorController("Shooter Motor", Constants.kShooterMotorPort, Constants.kShooterMotorCurrentLimit);
+    mHoodDoubleSolenoid = new DoubleSolenoid(Constants.kHoodDoubleSolenoidForwardChannel, Constants.kHoodDoubleSolenoidReverseChannel);
+    mShooterPID = mShooterMotor.getSparkMax().getPIDController();
+    mShooterPID.setP(mP);
+    mShooterPID.setI(mI);
+    mShooterPID.setD(mD);
+    SmartDashboard.putNumber("Shooter P Value", mP);
+    SmartDashboard.putNumber("Shooter I Value", mI);
+    SmartDashboard.putNumber("Shooter D Value", mD);
+  }
+
+  // Value.kReverse is when the hood is extended
+  public void setHoodExtended(boolean isExtended) {
+    if (isExtended)
+      mHoodDoubleSolenoid.set(Value.kReverse);
+    else  
+      mHoodDoubleSolenoid.set(Value.kForward);
+  }
+  
+  public boolean getHoodExtendRequired() {
+    return getDistanceFromGoal() > Constants.kHoodExtendRequiredDistance;
+  }
+
+  public void setVelocity(double velocity) {
+    mShooterVelocitySetpoint = velocity;
+    mShooterPID.setReference(velocity, ControlType.kVelocity);
+  }
+
+  public boolean getMotorVelocityWithinRange() {
+    // Compare if the motor velocity is within Constants.kShooterVelocityRange range
+    return (Math.abs(mShooterVelocitySetpoint - mShooterMotor.getEncoder().getVelocity()) < mShooterVelocitySetpoint * Constants.kShooterVelocityRange);
+  }
+
+  public boolean getRobotDistanceWithinRange() {
+    return getDistanceFromGoal() < getDistanceLimit();
+  }
+
+  public double getDistanceFromGoal() {
+    //Todo: get distance from limelight
+    return SmartDashboard.getNumber("Distance from Goal", 0.0);
+  }
+
+  public double getAngleFromGoal() {
+    //Todo: get angle from limelight
+    return SmartDashboard.getNumber("Angle from Goal", 0.0);
+  }
+
+  public boolean getAngleAligned() {
+    return getAngleFromGoal() == 0;
+  }
+
+  public boolean getReadyToShoot() {
+    return (getMotorVelocityWithinRange() && getAngleAligned());
+  }
+
+  public double getRequiredVelocityForDistance() {
+    double velocity; 
+    if (Constants.kGreenShootingZone < getDistanceFromGoal()) {
+      velocity = mGreenShootingVelocity;
+    } 
+    else if (Constants.kYellowShootingZone < getDistanceFromGoal()) {
+      velocity = mYellowShootingVelocity;
+    } 
+    else if (Constants.kBlueShootingZone < getDistanceFromGoal()) {
+      velocity = mBlueShootingVelocity;
+    }
+    else if (Constants.kRedShootingZone < getDistanceFromGoal()) {
+      velocity = mRedShootingVelocity;
+    }
+    else {
+      velocity = 0.0; //Todo: fix
```

Comment:
Fix?

Reply:
https://github.com/AusTINCANsProgrammingTeam/2021RobotCode/issues/13#issue-805057313 Is an issue to address this